New dependency tracking

eikenb opened the pull request:
Reworks how it keeps track of dependencies/views, templates and their relationships like which dependencies are used by which templates and visa versa.

2 main reasons for this.
1. to many objects with overlapping responsibilities, this cuts things down from ~6 objects to 1 (including sub-structs, ~10 to 3)
2. it eliminates the shared tracking of changes and localizes that to allow for per-template level checks

2 above will allow for per-template async management

Also includes some simplifications to the related interfaces as some methods were changed, added or made obsolete.
The resulting interfaces are smaller as well.

findkim commented on watcher.go:
```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
+
+// initialized returns true if the view has had its data fetched at least once
+func (t *tracker) initialized(viewID string) bool {
+	return t.views[viewID].receivedData
```
Does this work as intended? it reads to me that if `viewID` does not exist in the map, it'll be calling `.receivedData` on `nil`

https://play.golang.com/p/5TCcf3QUlKm

eikenb replied:
Yep. In the original version `views` was a `map[string]view` (non-pointer) which worked fine with the zero value. But then I changed it to `map[string]*view` which broke that. I really don't like pointer's zero value semantics at times.

findkim commented on watcher.go:
```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
+func (t *tracker) complete(n Notifier) bool {
+	for _, tp := range t.tracked {
+		thisNotifier := tp.notify == n.ID()
+		//		fmt.Println(thisNotifier, tp.inUse, !t.initialized(tp.view))
+		//		fmt.Printf("%#v\n", tp)
```
Cleanup!

eikenb replied:
I had left these here on purpose as I kept re-adding them while debugging... but I'll be optimistic and remove them for now.

findkim commented on watcher.go:
```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
+	return results
+}
+func (n *dummyNotifier) drain() {
+	for range n.deps {
+		// just draining
+	}
```
What's the purpose of `drain()`? This blocks without a way to stopping the function without the caller closing the channel.

```golang
close(n.deps)
```

eikenb replied:
Drain should have been removed. It isn't used and that method isn't even right.

findkim reviewed (commented):
The summary is underwhelming to the changes you made! This is a huge improvement:
* Notifies all templates watching a dependency when that dep has updated
* Reduces in-memory redundant caches

Not a request for this PR, but are there changes to the `doc_test.go` with these updates or a new example that these improvements have to offer?

eikenb commented:
> The summary is underwhelming to the changes you made! This is a huge improvement:
> 
>     * Notifies all templates watching a dependency when that dep has updated
> 
>     * Reduces in-memory redundant caches
> 
> 
> Not a request for this PR, but are there changes to the `doc_test.go` with these updates or a new example that these improvements have to offer?

I almost asked specifically for feedback on the commit message summary as I felt it was lacking but was to tired to come up with more. I think I'll steal these to add to it. Thanks!

eikenb commented on template.go:
```diff
@@ -6,12 +6,14 @@ import (
-var errTemplateMissingContents = errors.New("template: missing 'contents'")
+// ErrMissingValues is the error returned when a template doesn't completely
+// render due to missing values (values that haven't been fetched yet).
+var ErrMissingValues = errors.New("Missing template values")
+var ErrNoNewValues = errors.New("No new values for template")
```
Oh yeah... hard to overcome years of habit. Fixing.

eikenb commented on template.go:
```diff
@@ -167,51 +182,50 @@ func (t *Template) Execute(r Recaller) (*ExecuteResult, error) {
 	}
 
+	// Checks if all values in use have been fetched.
+	// ..also cleans out data no longer used by this template.
```
Aww. I like ellipsis. :disappointed: 
I guess.

eikenb commented on template.go:
```diff
@@ -167,51 +182,50 @@ func (t *Template) Execute(r Recaller) (*ExecuteResult, error) {
-		Used:    *used,
-		Missing: *missing,
-		Output:  b.Bytes(),
+		Output: b.Bytes(),
```
I thought about that as I was cleaning out the other stuff but was in the middle of things and wasn't sure. I'll make this change and see how it looks.

eikenb replied:
Required changes were spread out a bit, but not to bad. I'll push it up along with the rest (soon).

eikenb replied:
Here's the commit with them... I think this is good to go, but for context and in case.

https://github.com/hashicorp/hcat/pull/26/commits/9842aa03006bb6e422e0f073f527cc2ca827dc3e

eikenb commented on watcher.go:
```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
-	sync.RWMutex
-	deps depmap
-	tpls depmap
+	sync.Mutex
```
Good catch. I forget sometimes how embedding exposes its methods as public.

eikenb replied:
I see why I left it. I tend not to worry about public/private on structs that are already internal/private.

I mean giving it a private name changes nothing in terms of access as anything with access to Watcher.tracker would also have access to Wather.tracker.mux. So changing it would not impact the ability to accidentally call locks. I guess it would make it a touch harder.

:thinking:

lornasong replied:
@eikenb and I chatted about this offline and decided to keep the lock implementation as is. As mentioned in the previous comment, since the struct is private, public access to the lock is still prohibited. If the struct was public, we would definitely want to switch to a private lock. This implementation also makes calling the lock a little more elegant. But, if there's future concerns, we can update it later without any problems since it's private.

eikenb commented on watcher.go:
```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
 
-func (t *tracker) update(tmplID string, deps ...dep.Dependency) {
-	t.clear(tmplID)
+// lookup returns the view and true, or view's zero value and false
```
Another leftover of pre-pointer views map. Updated.

eikenb commented:
@kim I just pushed up a batch of changes addressing all the issues you found. I left them as separate commits for now, but will probably(?) squash them pre-merge.

eikenb commented on watcher.go:
```diff
@@ -177,13 +166,12 @@ func (w *Watcher) WaitCh(ctx context.Context) <-chan error {
-	cleanStop := make(chan struct{})
-	defer close(cleanStop) // only run while waiting
+	//	sweepStop := make(chan struct{})
+	//	defer close(sweepStop) // only run while waiting
 	go func() {
-		w.cleanDeps(cleanStop)
+		//w.tracker.sweep()
```
Yep. That's leftovers I missed cleaning out. Thanks.

eikenb commented on watcher.go:
```diff
@@ -196,7 +184,9 @@ func (w *Watcher) Wait(ctx context.Context) error {
 		id := v.Dependency().String()
 		w.cache.Save(id, v.Data())
-		w.changed.Add(id)
+		for _, n := range w.tracker.notifiersFor(v) {
+			n.Notify(v.Dependency())
+		}
```
Thanks!

lornasong reviewed (approved):
Cool refactor! Looking forward to seeing this play out in CTS. Just picking up findkim’s review where she left off since she’s out. Looks like previous comments are addressed and this looks good to me :tada:
